### DISCUSSION TITLE
Best way to implement endless canvas?

### DISCUSSION THREAD

**Discussion by lanyusan** [pixijs/pixijs#7957] [Q&A]
I am evaluating canvas libraries for a project to implement some sort of virtual canvas with endless size both vertical and horizontal.

I am asking if pixi has some best practices or built in mechanisms to support it.

Basically it is a virtual canvas with its own coordinate system and it could grow vertically and also horizontally endlessly. The actual screen is considered a sliding window to display a portion of the virtual canvas. A sliding window position, size and scale data structure would be maintained to decide what of the virtual canvas would be mapped to physical canvas and rendered.

Is there a best practice or Pixi way to do it and to make scrolling performant? 

Apology if it is too generic and conceptual. If you are familiar with Figma, basically the idea is the same to Figma's canvas, which could be extended endlessly by further scrolling when it reaches edges or drag canvas object further towards edges.

Thanks in advance.

**Comment by ivanpopelyshev** [comment]
https://www.pixiplayground.com/#/edit/B8R6ZZD8JROtw1c52LFnN

PixiJS is a low-level solution, for scrolling you also need good high-level algo , that you will write on your own. 

If you dont know math and culling for inifite things, if you didnt ever write 2d or 3d infinite games - it'll be a problem for you.

If you know how to do those things, you can go raise 10m$ for another Figma, like a few other projects on pixi did.

**Reply by lanyusan** [reply]
well i happen to be math guy. :) 
But i haven't done infinite things before. I get it would be challenging.
By the way, is Figma using Pixi or some other library or an in house one?

**Reply by ivanpopelyshev** [reply]
homebrew, with lots of canvas2d usage and caching , and their own blur filter

**Reply by lanyusan** [reply]
Thanks

**Reply by photonstorm** [reply]
> homebrew, with lots of canvas2d usage and caching , and their own blur filter

And a very big and well funded development team 🙂
